lib/puppet/configurer.rb

```diff
@@ -126,6 +126,73 @@ def convert_catalog(result, duration, facts, options = {})
     catalog
   end
 
+  def warn_name_length(name, length, max_length)
+    Puppet.warning _("Fact %{name} with length: '%{length}' exceedes the length limit: %{limit}") % { name: name, length: length, limit: max_length }
+  end
+
+  def warn_size_length(size, max_size)
+    Puppet.warning _("The current fact size: %{size} exceeds the number of top facts limit: %{max_size}") % { size: size, max_size: max_size }
+  end
+
+  def warn_value_length(key, value, length, max_length)
+    Puppet.warning _("Fact '%{key}' with value '%{value}' and the value length: '%{length}' exceeds the value length limit: %{max_length}") % { key: key, value: value, length:length, max_length: max_length }
+  end
+
+  def check_fact_name_length(name)
+    max_length = Puppet[:max_factnamelength]
+    return if max_length == -1
+
+    max_length = 100 if max_length == 0
+    warn_name_length(name, name.length, max_length) if name.bytesize > max_length
+  end
+
+  def check_fact_values_length(keys,values)
+    max_length = Puppet[:max_factvaluelength]
+    return if max_length == -1
+
+    max_length = 512 if max_length == 0
+    warn_value_length(keys, values, values.to_s.length, max_length) if values.to_s.bytesize > max_length
+  end
+
+  def check_size_limits(size)
+    max_size = Puppet[:max_toplevelfacts]
+    return if max_size == -1
+
+    max_size = 200 if max_size == 0
+    warn_size_length(size, max_size) if size > max_size
+  end
+
+  def get_fact_values(keys, values)
+    if values.is_a?(Hash) 
+      values.each do |k,v|
+        value = v || k
+        if value.is_a?(Hash) || value.is_a?(Array)
+          get_fact_values(keys, value)
+        else
+          check_fact_values_length(k, v)
+        end
+      end
+    elsif values.is_a?(Array)
+      values.each do |value|
+        if value.is_a?(Hash) || value.is_a?(Array)
+          get_fact_values(keys, value)
+        else
+          check_fact_values_length(keys, value)
+        end
+      end
+    else
+      check_fact_values_length(keys, values)
+    end
+  end
+
+  def check_top_level(facts)
```

Comment:
This is called `check_top_level` but it actually iterates inside the facts hash and checks everything